### ISSUE TITLE
laggy clients makes nginx core dump in nchan

### ISSUE THREAD

**Issue by DomrachevIvan** [slact/nchan#396]
if laggy client do not close connection and do not read from socket - nchan cure dump.
there are a lot of different error logs:
```
2017/09/27 14:52:57 [error] 13337#100108: MEMSTORE:07: tried adding WAITING chanhead 0000000802406800 / to chanhead_gc. why?
Assertion failed: (data.shm_chid->len>1), function memstore_ipc_send_get_message, file /home/user/current/ports/10/2016.05.31/nginx-cache/app/work/nchan-1.1.7/src/store/memory/ipc-handlers.c, line 462.
2017/09/27 15:02:12 [alert] 13608#100110: worker process 13611 exited on signal 6 (core dumped)
2017/09/27 15:02:45 [error] 14715#100069: MEMSTORE:02: tried adding WAITING chanhead 0000000802406800 / to chanhead_gc. why?
2017/09/27 15:02:45 [error] 13613#100079: MEMSTORE:04: force-reaping msg with refcount 1
Assertion failed: (data.shm_chid->len>1), function memstore_ipc_send_get_message, file /home/user/current/ports/10/2016.05.31/nginx-cache/app/work/nchan-1.1.7/src/store/memory/ipc-handlers.c, line 462.
2017/09/27 15:29:40 [alert] 16685#100100: worker process 16693 exited on signal 6 (core dumped)
2017/09/27 15:29:42 [alert] 16685#100100: worker process 16732 exited on signal 11 (core dumped)
```
reproduce steps:
start nginx
start poster
start laggy client
wait for 1 minute or 5.
sometimes nginx process core dump till test, sometimes it just hangs and do not reply on requests (but if you try to stop it, it will core dump).
it doesn't mater nchan_message_buffer_length 0 or 5000. coredump anyway.

os: freebsd 10.3
nginx config:
```nginx
worker_processes 8;
error_log shared/nginx/logs-raw/e.current warn;
pid shared/nginx/pid;

events {
 worker_connections 10240;
}

http {

 client_body_temp_path shared/nginx/temp/client-body;
 proxy_temp_path shared/nginx/temp/proxy;
 fastcgi_temp_path shared/nginx/temp/fastcgi;
 uwsgi_temp_path shared/nginx/temp/uwsgi;
 scgi_temp_path shared/nginx/temp/scgi;

 access_log off;

 server {
  listen *:80;
  server_name localhost;

  location /live/ {
   nchan_store_messages off;
   nchan_message_buffer_length 0;
   nchan_subscriber_timeout 30s;
   nchan_subscriber_first_message newest;

   location /live/sub/index {
    nchan_subscriber eventsource;
    nchan_channel_id $1;
   }
   location /live/pub/index {
    nchan_publisher;
    nchan_channel_id $1;
   }
   location /live/status {
    nchan_stub_status;
   }
  }
 }
}
```

poster:
```perl
#!/usr/bin/env perl

use LWP::UserAgent;

$ua=LWP::UserAgent->new;

for(;;){
 foreach(1..400){
  $response=$ua->post("http://localhost/live/pub/index",Content=>"{".('a'x(1000*rand))."}");
  print $response->content;
  print "\n\n";
 }
 sleep(2);
}
```

laggy client:
```perl
#!/usr/bin/env perl
use Socket;
for(;;){
 print ".\n";
 socket(S,PF_INET,SOCK_STREAM,getprotobyname("tcp")) || die "socket: $!";
 connect(S,pack_sockaddr_in(80,inet_aton('127.0.0.1'))) || die "connect: $!";
 send(S,<<EOF,0);
GET /live/sub/index HTTP/1.1
Host: localhost
Connection: Close
Accept: text/event-stream

EOF
 <S>;
 sleep(30);
 close(S);
}
```

**Comment by slact** [comment]
What version of nginx and nchan are you using?

**Comment by DomrachevIvan** [comment]
nginx 1.13.4
nchan 1.1.7

**Comment by slact** [comment]
Unable to reproduce after running your provided scripts with the config given for a few hours. 

Could I perhaps see what's happening on your machine?

**Comment by DomrachevIvan** [comment]
sent email with link to vmware virtual machine with freebsd 10.3 and prepared nginx.

the same core dumps with nchan_max_reserved_memory 512M

**Comment by slact** [comment]
Thhanks, I can reproduce it now. Will update when I have a fix.

**Comment by slact** [comment]
Ok, easy fix. The issue was a channel id length of 0, from the `nchan_channel_id $1; ` setting, which evaluates to `""`.
The laggy client didn't have much to do with it.

Fixed in master in commit 90134c63.

**Comment by DomrachevIvan** [comment]
it's great!
but it was another bug. mine bug. till minifyng config and scripts I lost regexp param.
but coredumps are still here.
new config:
```
worker_processes 8;
error_log /var/log/nginx/error.log  warn;

events {
 worker_connections 10240;
}

http {

 access_log off;
 nchan_max_reserved_memory 256M;

 server {
  listen *:80;
  server_name localhost;

  location /live/ {
   nchan_store_messages off;
   nchan_message_buffer_length 0;
   nchan_subscriber_timeout 30s;
   nchan_subscriber_first_message newest;

   location /live/sub/index {
    nchan_subscriber eventsource;
    nchan_channel_id index;
   }
   location /live/pub/index {
    nchan_publisher;
    nchan_channel_id index;
   }
   location /live/status {
    nchan_stub_status;
   }
  }
 }
}
```
and new client:
```
#!/usr/bin/env perl

use Socket;
$SIG{PIPE}='IGNORE';
for(;;){
 my @h=();
 foreach(1..500){
   my $s;
   socket($s,PF_INET,SOCK_STREAM,getprotobyname("tcp"));
   connect($s,pack_sockaddr_in(80,inet_aton('127.0.0.1')));
 send($s,<<EOF,0);
GET /live/sub/index HTTP/1.1
Host: localhost
Connection: Close
Accept: text/event-stream

EOF
   push(@h,$s);
  }
 foreach(@h){<$_>}
 sleep(30);
 print "\n";
}
```

**Comment by prsephton** [comment]
+1 also having this issue.  Works for a whole, then barfs when clients are laggy.

**Comment by slact** [comment]
@prsephton : what is your total message publishing rate?

**Comment by prsephton** [comment]
Hi, Leo,

I don't believe it is related to publishing rate.

In my instance,  the core dump can happen at any time of day or night;  For
example, it happened at 1:29 AM where none of my five channels were active,
other than for the 'test' channel which publishes once every 5s.  I can
confirm that nothing was listening on the 'test' channel at this time, and
that the channel had no connections prior to it failing.

It is as if the channel simply runs out of memory, even though internal
queue size should be only around 10 items.  Once a failure occurs,
nginx/nchan remains in a broken state until nginx is restarted.  Simply
restarting the publishing app is not good enough.

Even in a failed state, nginx itself continues to function and process
requests, although the nchan part is inactive.

The core dump happens on the same assertion failure as previously reported:

/src/store/memory/ipc-handlers.c:434: memstore_ipc_send_get_message:
Assertion ` data.shm_chid->len>1' failed.

Looking at the code seems to indicate a shared memory resource limit issue
since the shared memory copy clearly failed.

I get other unexplained errors:
2017/10/05 11:09:46 [error] 9089#0: MEMSTORE:00: tried adding WAITING
chanhead 0
9D10650 /Activity to chanhead_gc. why?
2017/10/05 11:11:04 [error] 9089#0: MEMSTORE:00: tried adding WAITING
chanhead 0
9D12050 /HallOfFame.daily.school to chanhead_gc. why?
2017/10/05 11:11:09 [error] 9089#0: MEMSTORE:00: tried adding WAITING
chanhead 0
9D1D518 /HallOfFame.weekly.school to chanhead_gc. why?
2017/10/05 11:11:42 [error] 9089#0: MEMSTORE:00: tried adding WAITING
chanhead 0
9D11C20 /HallOfFame.weekly.teacher to chanhead_gc. why?
2017/10/05 12:25:45 [error] 9090#0: MEMSTORE:01: tried adding WAITING
chanhead 0
9DCDAF8 /HallOfFame.daily.student to chanhead_gc. why?
2017/10/05 12:25:46 [error] 9090#0: MEMSTORE:01: tried adding WAITING
chanhead 09DCE2A8 /HallOfFame.daily.school to chanhead_gc. why?
2017/10/05 12:32:58 [error] 9091#0: MEMSTORE:02: tried adding WAITING
chanhead 09CFF288 /HallOfFame.daily.school to chanhead_gc. why?
2017/10/05 12:33:33 [error] 9091#0: MEMSTORE:02: tried adding WAITING
chanhead 09D1D230 /HallOfFame.monthly.school to chanhead_gc. why?
2017/10/05 12:34:15 [error] 9091#0: MEMSTORE:02: tried adding WAITING
chanhead 09CBBF00 /Activity to chanhead_gc. why?
2017/10/05 12:35:07 [error] 9091#0: MEMSTORE:02: tried adding WAITING
chanhead 09CBBF00 /Activity to chanhead_gc. why?
... etc ...

These were similar to the ones reported by the other poster

I don't know what these errors mean.

I hope this helps!


<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by slact** [comment]
@prsephton : any chance I could take a look at the coredumps?

**Comment by prsephton** [comment]
Sure, I don't have a problem with that.  Where would I find the core dump?
From what I recall, the environment that nginx runs in must be specifically
set up to dump the core to file, not so?

I did a quick find / -iname '*.core', and that did not come up with
anything promising.

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by prsephton** [comment]
Oh yes, now I remember: ulimit -c unlimited

I will let you have the next one I get

**Comment by slact** [comment]
I need the first coredump after you restart (not just reload) nginx. You'll need to change your kernel's core_pattern paremeter to make this easy.

as root, run
echo "/path/to/coredumps/core.%t-%e.%p-%s" > /proc/sys/kernel/core_pattern

Make sure the directory exists, and you'll get nice coredump files that
look like 'core.1454680914-nginx.4053-11'

**Comment by prsephton** [comment]
thanks!


<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by kilgor** [comment]
To have full core dump you should add to nginx .conf
worker_rlimit_core  3000M;
working_directory   /path/to/coredumps/;

And as @slact said modify proc/sys/kernel/core_pattern

**Comment by prsephton** [comment]
Just a little feedback: I have not forgotten, just no core dumps yet.  Sometimes it can go days without crashing.

**Comment by prsephton** [comment]
Finally, I had a core dump last night.  You can access it at [core.1508272415-nginx.28777-6.bz2](www.cami-apps.com/files/core.1508272415-nginx.28777-6.bz2).

Let me know if there is anything else I can help with.

Kind regards

**Comment by DomrachevIvan** [comment]
just tested v1.1.8.
still got coredumps, hangs and mem leak with my scripts.
what can I do to help you find out the problem?

**Comment by slact** [comment]
sorry, been very busy past two weeks. I will resume this thread on Tuesday.

**Comment by DomrachevIvan** [comment]
thank you

**Comment by slact** [comment]
@DomrachevIvan : I would like to take a look and maybe do some testing on your server. contact me at leo@nchan.io to arrange that.

**Comment by slact** [comment]
fixed in 1.1.10
